### PR TITLE
[SPARK-34104][SPARK-34105][CORE][K8S] Maximum decommissioning time & allow decommissioning for excludes

### PR DESCRIPTION
**Pull request by holdenk**
### What changes were proposed in this pull request?

Allow users to have Spark attempt to decommission excluded executors. 
Since excluded executors may be flaky, this also adds the ability for users to specify a time limit after which a decommissioning executor will be killed by Spark.

### Why are the changes needed?

This may help prevent fetch failures from excluded executors, and also handle the situation in which executors 

### Does this PR introduce _any_ user-facing change?

Yes, two new configuration flags for the behaviour.

### How was this patch tested?

Extended unit and integration tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @Ngone51 FYI

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 1]
```diff
@@ -1933,7 +1939,8 @@ package object config {
 
   private[spark] val EXECUTOR_DECOMMISSION_KILL_INTERVAL =
     ConfigBuilder("spark.executor.decommission.killInterval")
-      .doc("Duration after which a decommissioned executor will be killed forcefully." +
+      .doc("Duration after which a decommissioned executor will be killed forcefully " +
+        " *by an outside* (e.g. non-spark) service. " +
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1943]: nit. Remove the leading space, `" *by` -> `"*by`.

File: core/src/main/scala/org/apache/spark/scheduler/HealthTracker.scala [hunk 2]
```diff
@@ -157,8 +159,13 @@ private[scheduler] class HealthTracker (
     allocationClient match {
       case Some(a) =>
         logInfo(msg)
-        a.killExecutors(Seq(exec), adjustTargetNumExecutors = false, countFailures = false,
-          force = true)
+        if (!decommission) {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/scheduler/HealthTracker.scala:162]: Instead of `negative if` statement, shall we simplify by switch the block (line 166~167) and the block (line 163 ~ 164)?

File: core/src/main/scala/org/apache/spark/scheduler/HealthTracker.scala [hunk 3]
```diff
@@ -184,8 +191,14 @@ private[scheduler] class HealthTracker (
         case Some(a) =>
           logInfo(s"Killing all executors on excluded host $node " +
             s"since ${config.EXCLUDE_ON_FAILURE_KILL_ENABLED.key} is set.")
-          if (a.killExecutorsOnHost(node) == false) {
-            logError(s"Killing executors on node $node failed.")
+          if (!decommission) {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/scheduler/HealthTracker.scala:194]: ditto.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationClient.scala [hunk 4]
```diff
@@ -130,6 +130,13 @@ private[spark] trait ExecutorAllocationClient {
   }
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationClient.scala:132]: nit. This is irrelevant, but can we remove this empty line 132?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 5]
```diff
@@ -115,6 +115,12 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
   private val reviveThread =
     ThreadUtils.newDaemonSingleThreadScheduledExecutor("driver-revive-thread")
 
+  private val cleanupService: Option[ScheduledExecutorService] =
+    conf.get(EXECUTOR_DECOMMISSION_CLEANUP_INTERVAL).map { _ =>
+      ThreadUtils.newDaemonSingleThreadScheduledExecutor("cleanup-decommission-execs")
+    }
+
+
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:123]: nit. Shall we remove this empty line?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 6]
```diff
@@ -506,6 +522,21 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
       }
     }
 
+    conf.get(EXECUTOR_DECOMMISSION_CLEANUP_INTERVAL).map {
+      cleanup_interval =>
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:526]: If you don't mind, shall we move this to the upper line like the following?
```scala
conf.get(EXECUTOR_DECOMMISSION_CLEANUP_INTERVAL).map { cleanup_interval =>
```

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:526]: And, maybe, the variable name `cleanup_interval` looks like Python convention. Shall we renaming to `cleanupInterval` or `internal` simply?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 7]
```diff
@@ -506,6 +522,21 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
...
+      val cleanupTask = new Runnable() {
+        override def run(): Unit = Utils.tryLogNonFatalError {
+          val stragglers = executorsToDecommission.filter(executorsPendingDecommission.contains(_))
+          if (!stragglers.isEmpty) {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:530]: Maybe, the following?
```scala
- if (!stragglers.isEmpty) {
+ if (stragglers.nonEmpty) {
```

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 8]
```diff
@@ -850,6 +882,22 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
   protected def doKillExecutors(executorIds: Seq[String]): Future[Boolean] =
     Future.successful(false)
 
+  /**
+   * Request that the cluster manager decommission all executors on a given host.
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:886]: `decommission` -> `decommissions`

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 9]
```diff
@@ -850,6 +882,22 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
...
+   * @return whether the decommission request is acknowledged.
+   */
+  final override def decommissionExecutorsOnHost(host: String): Boolean = {
+    logInfo(s"Requesting to kill any and all executors on host ${host}")
+    // A potential race exists if a new executor attempts to register on a host
+    // that is on the exclude list and is no no longer valid. To avoid this race,
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:892]: `no no longer` -> `no longer`

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala [hunk 10]
```diff
@@ -345,8 +345,11 @@ class KubernetesSuite extends SparkFunSuite
                 }
                 // Delete the pod to simulate cluster scale down/migration.
                 // This will allow the pod to remain up for the grace period
+                // We set an intentionally long grace period to test that Spark
+                // exits once the blocks are done migrating and doesn't wait for the
+                // entire grace period if it does not need to.
                 kubernetesTestComponents.kubernetesClient.pods()
-                  .withName(name).delete()
+                  .withName(name).withGracePeriod(8640000).delete()
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:352]: Although this is a test case, `8640000` looks like a magic number. Is there a reason to choose this?

Comment (dongjoon-hyun) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:352]: Is this due to https://github.com/apache/spark/pull/31249/files#diff-1c2fb0706df8a5c08f357f108f907239a945d3b857c3d2a774c59fc083b57416R130 ?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 11]
```diff
@@ -176,11 +182,21 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
         }
 
       case KillExecutorsOnHost(host) =>
-        scheduler.getExecutorsAliveOnHost(host).foreach { exec =>
-          killExecutors(exec.toSeq, adjustTargetNumExecutors = false, countFailures = false,
+        scheduler.getExecutorsAliveOnHost(host).foreach { execs =>
+          killExecutors(execs.toSeq, adjustTargetNumExecutors = false, countFailures = false,
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:186]: If you don't mind, shall we revert this change to reduce this patch size?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 12]
```diff
@@ -506,6 +522,21 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
       }
     }
 
+    conf.get(EXECUTOR_DECOMMISSION_CLEANUP_INTERVAL).map {
+      cleanup_interval =>
+      val cleanupTask = new Runnable() {
+        override def run(): Unit = Utils.tryLogNonFatalError {
+          val stragglers = executorsToDecommission.filter(executorsPendingDecommission.contains(_))
+          if (!stragglers.isEmpty) {
+            logError(
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:531]: Decommission is not guaranteed always by design and it's just our best effort approach. In addition, we delete it in any way at line 533. Shall we make Spark show `logWarn` instead of `logError`?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 13]
```diff
@@ -850,6 +882,22 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
   protected def doKillExecutors(executorIds: Seq[String]): Future[Boolean] =
     Future.successful(false)
 
+  /**
+   * Request that the cluster manager decommission all executors on a given host.
+   * @return whether the decommission request is acknowledged.
+   */
+  final override def decommissionExecutorsOnHost(host: String): Boolean = {
+    logInfo(s"Requesting to kill any and all executors on host ${host}")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:890]: nit. `${host}` -> `$host`?

File: core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala [hunk 14]
```diff
@@ -554,6 +554,49 @@ class HealthTrackerSuite extends SparkFunSuite with BeforeAndAfterEach with Mock
...
+    val allocationClientMock = mock[ExecutorAllocationClient]
+    healthTracker = new HealthTracker(listenerBusMock, conf, Some(allocationClientMock), clock)
+
+    // verify we decommission when configured
+    conf.set(config.EXCLUDE_ON_FAILURE_KILL_ENABLED, true)
+    conf.set(config.DECOMMISSION_ENABLED.key, "true")
+    conf.set(config.EXCLUDE_ON_FAILURE_DECOMMISSION_ENABLED.key, "true")
+    healthTracker = new HealthTracker(listenerBusMock, conf, Some(allocationClientMock), clock)
+
+    // Fail 4 tasks in one task set on executor 1, so that executor gets excluded for the whole
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala:567]: This seems to depend on our default configuration. Instead of assuming, can we use the default value of the confs directly instead?

File: core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala [hunk 15]
```diff
@@ -554,6 +554,49 @@ class HealthTrackerSuite extends SparkFunSuite with BeforeAndAfterEach with Mock
...
+    healthTracker.updateExcludedForSuccessfulTaskSet(0, 0, taskSetExclude2.execToFailures)
+
+    val msg1 =
+      "Killing excluded executor id 1 since spark.excludeOnFailure.killExcludedExecutors is set."
+
+    verify(allocationClientMock).decommissionExecutor(
+      "1", ExecutorDecommissionInfo(msg1), false)
+
+    val taskSetExclude3 = createTaskSetExcludelist(stageId = 1)
+    // Fail 4 tasks in one task set on executor 2, so that executor gets excluded for the whole
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala:583]: ditto.

File: core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala (hunk 14 above)
Comment (dongjoon-hyun) [reply core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala:567]: The default can be changed later in Apache Spark and can be different in the downstream, @holdenk .

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you, @holdenk . Mostly, looks good. I added a few comments.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala (hunk 10 above)
Comment (holdenk) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:352]: This is just a really long value, we can pick any large number here.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 11 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:186]: So I switched the name to `execs` since it is multiple executors and I was having difficulty reading the code the first time through. It's not strong feelings but I think it's clearer this way (e.g. it's not calling foreach on the executors it's actually calling foreach on the option of list of execs).

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 12 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:531]: Yeah good point, I think info might actually be better.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 13 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:890]: this matches the killExecsonHost code, so I'd either change both or not.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala (hunk 10 above)
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:352]: In that case, can we use `Int.MaxValue` simply?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 11 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:185]: Got it. It's up to your decision.

Comment (holdenk) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:185]: cool, if your ok with it I find this makes the code more readable so I'll leave as is :)

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 16]
```diff
@@ -809,6 +809,12 @@ package object config {
       .booleanConf
       .createWithDefault(false)
 
+  private[spark] val EXCLUDE_ON_FAILURE_DECOMMISSION_ENABLED =
+    ConfigBuilder("spark.excludeOnFailure.decommissionExcludedExecutors")
+      .version("3.2.0")
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:814]: BTW, can we add docs since it's an exposed configuration?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 17]
```diff
@@ -506,6 +521,20 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
       }
     }
 
+    conf.get(EXECUTOR_DECOMMISSION_CLEANUP_INTERVAL).map { cleanupInterval =>
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:524]: IIUC, the decommissioned executor is supposed to shutdown et the end whether by the external services or itself (when decommissioned is considered to be done), right?

I understand this clean task is a guarantee for executor shutdown. But I think, for users, it's really hard to get a reasonable value of the conf here since they don't know what happens during decommissioning and what the approximate time it would take.

Besides, in the case of decommission coming from the external services, I think we'd better respect the `spark.executor.decommission.killInterval` here. This would prevent us from killing the executor prematurely when `spark.executor.decommission.cleanupInterval` is shorter than `spark.executor.decommission.killInterval`.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 18]
```diff
@@ -839,7 +869,6 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
         Future.successful (if (killSuccessful) executorsToKill else Seq.empty[String])
       )(ThreadUtils.sameThread)
     }
-
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:842]: Could you revert this unrelated change?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 13 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:887]: +1 to change both

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:887]: also, "kill any executors" or "kill all executors" should be enough?

File: core/src/main/scala/org/apache/spark/scheduler/HealthTracker.scala [hunk 19]
```diff
@@ -40,6 +40,7 @@ import org.apache.spark.util.{Clock, SystemClock, Utils}
  *      stage, but still many failures over the entire application
  *  * "flaky" executors -- they don't fail every task, but are still faulty enough to merit
  *      excluding
+ *  * missing shuffle files -- may trigger fetch failures on health executors.
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/HealthTracker.scala:43]: nit: "healthy"?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 20]
```diff
@@ -506,6 +521,20 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
       }
     }
 
+    conf.get(EXECUTOR_DECOMMISSION_CLEANUP_INTERVAL).map { cleanupInterval =>
+      val cleanupTask = new Runnable() {
+        override def run(): Unit = Utils.tryLogNonFatalError {
+          val stragglers = executorsToDecommission.filter(executorsPendingDecommission.contains(_))
```
[the remark this answers is not in the document]
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:527]: Also, the task is executed in a separate thread so we should ensure the thread-safe on `executorsPendingDecommission`.

File: core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala [hunk 21]
```diff
@@ -554,6 +554,51 @@ class HealthTrackerSuite extends SparkFunSuite with BeforeAndAfterEach with Mock
     verify(allocationClientMock).killExecutorsOnHost("hostA")
   }
 
+  test("excluding decommission and kills executors when enabled") {
+    val allocationClientMock = mock[ExecutorAllocationClient]
+    healthTracker = new HealthTracker(listenerBusMock, conf, Some(allocationClientMock), clock)
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala:559]: unused initialization?

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 22]
```diff
@@ -809,6 +809,12 @@ package object config {
       .booleanConf
       .createWithDefault(false)
 
+  private[spark] val EXCLUDE_ON_FAILURE_DECOMMISSION_ENABLED =
+    ConfigBuilder("spark.excludeOnFailure.decommissionExcludedExecutors")
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:813]: Seems like this only takes effect when `spark.excludeOnFailure.killExcludedExecutors=true`. If so, I think it's more reasonable to have the same namespace for this conf, e.g., `spark.excludeOnFailure.killExcludedExecutors.decommissionEnabled` or something else.

And with the current implementation, the executor may not get killed if `EXECUTOR_DECOMMISSION_CLEANUP_INTERVAL` is not set, which then doesn't respect the `spark.excludeOnFailure.killExcludedExecutors=true` indeed. 

Actually, I'm thinking that it might be better to have these two confs at the same level?, e.g.,

```scala
if (EXCLUDE_ON_FAILURE_KILL_ENABLED) {
 // do kill
} else if (EXCLUDE_ON_FAILURE_DECOMMISSION_ENABLED) {
 // do decommission
}
// else do nothing
```

WDYT?

File: core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala [hunk 23]
```diff
@@ -554,6 +554,51 @@ class HealthTrackerSuite extends SparkFunSuite with BeforeAndAfterEach with Mock
...
+
+    // verify we decommission when configured
+    conf.set(config.EXCLUDE_ON_FAILURE_KILL_ENABLED, true)
+    conf.set(config.DECOMMISSION_ENABLED.key, "true")
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala:563]: this conf is not required?

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 22 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:813]: I like the nested one because if they're both set to true in the second configuration priority is ambiguous.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 17 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:523]: I think users should be able to understand that block migrations are occurring, or if they aren't they probably won't be customizing this part of the configuration anyways. I can add a note in the config docs that this should be a relatively high value.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 20 above)
[the remark this answers is not in the document]
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:526]: Good catch on the thread safety

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 13 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:886]: this seems like bike shedding on a log message that I'd rather avoid.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala (hunk 10 above)
Comment (holdenk) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:352]: Sounds good to me.

File: core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala (hunk 14 above)
Comment (holdenk) [reply core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala:568]: I've added explicit configs.

File: core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala (hunk 23 above)
Comment (holdenk) [reply core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala:562]: Without decommissioning enabled the executor _could_ do something different (although right now your right this is not needed). It just seemed like a better future proof configuration to have it on.

**Comment by holdenk** [comment]
I think I've addressed all of the outstanding issues, if no one has any remaining concerns I'll merge this tomorrow.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 24]
```diff
@@ -1942,6 +1950,15 @@ package object config {
       .timeConf(TimeUnit.SECONDS)
       .createOptional
 
+  private[spark] val EXECUTOR_DECOMMISSION_CLEANUP_INTERVAL =
+    ConfigBuilder("spark.executor.decommission.cleanupInterval")
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1954]: Shall we rename it to `*.decommission.forceKillTimeout`? `Interval` sounds more like a periodical task while it actually happens once.

File: core/src/main/scala/org/apache/spark/scheduler/HealthTracker.scala [hunk 25]
```diff
@@ -64,6 +65,7 @@ private[scheduler] class HealthTracker (
   val EXCLUDE_ON_FAILURE_TIMEOUT_MILLIS = HealthTracker.getExludeOnFailureTimeout(conf)
   private val EXCLUDE_FETCH_FAILURE_ENABLED =
     conf.get(config.EXCLUDE_ON_FAILURE_FETCH_FAILURE_ENABLED)
+  private val decommission = conf.get(config.EXCLUDE_ON_FAILURE_DECOMMISSION_ENABLED)
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/HealthTracker.scala:68]: Shall we keep the same pattern of naming here? e.g., `EXCLUDE_ON_FAILURE_DECOMMISSION_ENABLED`. Usually, it's good for developers to understand the code here.

File: core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala [hunk 26]
```diff
@@ -554,6 +554,50 @@ class HealthTrackerSuite extends SparkFunSuite with BeforeAndAfterEach with Mock
...
+    // application.
+    val taskSetExclude2 = createTaskSetExcludelist(stageId = 0)
+    (0 until 4).foreach { partition =>
+      taskSetExclude2.updateExcludedForFailedTask(
+        "hostA", exec = "1", index = partition, failureReason = "testing")
+    }
+    healthTracker.updateExcludedForSuccessfulTaskSet(0, 0, taskSetExclude2.execToFailures)
+
+    val msg1 =
+      "Killing excluded executor id 1 since spark.excludeOnFailure.killExcludedExecutors is set."
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala:578]: Shall we include decommission hint when enabled in the message (in `killExecutor()`)? e.g., "Killing (decommission) excluded executor..."

File: core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala [hunk 27]
```diff
@@ -554,6 +554,50 @@ class HealthTrackerSuite extends SparkFunSuite with BeforeAndAfterEach with Mock
...
+        "hostA", exec = "2", index = partition, failureReason = "testing")
+    }
+    healthTracker.updateExcludedForSuccessfulTaskSet(0, 0, taskSetExclude3.execToFailures)
+
+    val msg2 =
+      "Killing excluded executor id 2 since spark.excludeOnFailure.killExcludedExecutors is set."
+    verify(allocationClientMock).decommissionExecutor(
+      "2", ExecutorDecommissionInfo(msg2), false, false)
+    verify(allocationClientMock).decommissionExecutorsOnHost(
+      "hostA")
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala:598]: nit: turn into one line?

**Comment by Ngone51** [comment]
I left some minor comments. Overall, looks good to me.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 24 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:1954]: Yeah sure that sounds like a reasonable change.

File: core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala (hunk 26 above)
Comment (holdenk) [reply core/src/test/scala/org/apache/spark/scheduler/HealthTrackerSuite.scala:578]: sure

**Comment by holdenk** [comment]
Thanks @Ngone51 :)

**Comment by holdenk** [comment]
Merged to the current dev branch.

**Comment by HyukjinKwon** [comment]
Hey .. this broke Spark 2.13 build.. please avoid merging when CI did not pass..

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DecommissionSuite.scala [hunk 28]
```diff
@@ -116,6 +116,38 @@ private[spark] trait DecommissionSuite { k8sSuite: KubernetesSuite =>
...
+      .set("spark.kubernetes.container.image", pyImage)
+      .set(config.STORAGE_DECOMMISSION_ENABLED.key, "true")
+      .set(config.STORAGE_DECOMMISSION_SHUFFLE_BLOCKS_ENABLED.key, "true")
+      .set(config.STORAGE_DECOMMISSION_RDD_BLOCKS_ENABLED.key, "true")
+      // Ensure we have somewhere to migrate our data too
+      .set("spark.executor.instances", "3")
+      // Set super high so the timeout is triggered
+      .set("spark.storage.decommission.replicationReattemptInterval", "8640000")
+      // Set super low so the timeout is triggered
+      .set(config.EXECUTOR_DECOMMISSION_CLEANUP_INTERVAL.key, "10")
```
Comment (HyukjinKwon) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DecommissionSuite.scala:132]: `EXECUTOR_DECOMMISSION_CLEANUP_INTERVAL` this doesn't exists in the codes @holdenk. i will revert this one to recover the build.

**Comment by holdenk** [comment]
Sorry about that @HyukjinKwon , I thought we were going by Jenkins test pass on days when GH actions are flaky (https://twitter.com/githubstatus/status/1359176615931105282 ). I'm confused why it would have issues with 2.13 and build ok in Jenkins.

**Comment by holdenk** [comment]
I guess in Jenkins we don't build with the K8s profile which is why CI gave it a green.

**Comment by HyukjinKwon** [comment]
Yeah, it's fine. It happens. Thanks for a quick fix.

**Comment by HeartSaVioR** [comment]
> Kubernetes integration test status failure
> URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39633/

Probably we need to make sure this build is also happy if the PR is dealing with k8s.
